CodeConverter/CSharp/ExpressionNodeVisitor.cs

```diff
@@ -1818,8 +1818,54 @@ private ArgumentSyntax CreateExtraArgOrNull(IParameterSymbol p, bool requiresCom
     private ArgumentSyntax CreateOptionalRefArg(IParameterSymbol p, RefKind refKind)
     {
         string prefix = $"arg{p.Name}";
-        var local = _typeContext.PerScopeState.Hoist(new AdditionalDeclaration(prefix, CommonConversions.Literal(p.ExplicitDefaultValue), CommonConversions.GetTypeSyntax(p.Type)));
+        var type = CommonConversions.GetTypeSyntax(p.Type);
+        ExpressionSyntax initializer;
+        if (p.HasExplicitDefaultValue) {
+            initializer = CommonConversions.Literal(p.ExplicitDefaultValue);
+        } else if (HasOptionalAttribute(p)) {
+            if (TryGetDefaultParameterValueAttributeValue(p, out var defaultValue)){
+                initializer = CommonConversions.Literal(defaultValue);
+            } else {
+                initializer = SyntaxFactory.DefaultExpression(type);
+            }
+        } else {
+            //invalid VB.NET code
+            return null;
+        }
+        var local = _typeContext.PerScopeState.Hoist(new AdditionalDeclaration(prefix, initializer, type));
         return (ArgumentSyntax)CommonConversions.CsSyntaxGenerator.Argument(p.Name, refKind, local.IdentifierName);
+
+        bool HasOptionalAttribute(IParameterSymbol p)
+        {
+            var optionalAttribute = _semanticModel.Compilation.GetTypeByMetadataName("System.Runtime.InteropServices.OptionalAttribute");
+            if (optionalAttribute == null) {
+                return false;
+            }
+
+            return p.GetAttributes().Any(a => SymbolEqualityComparer.IncludeNullability.Equals(a.AttributeClass, optionalAttribute));
+        }
+    
+        bool TryGetDefaultParameterValueAttributeValue(IParameterSymbol p, out object defaultValue)
+        {
+            defaultValue = null;
+
+            var defaultParameterValueAttribute = _semanticModel.Compilation.GetTypeByMetadataName("System.Runtime.InteropServices.DefaultParameterValueAttribute");
```

Comment:
The implementation only caches the last 50 requested types, and they can be a bit expensive to find, so I tend to try to stash them in a field if they may be hit a lot like these.

Reply:
and what about using `.ToDisplayString(SymbolDisplayFormat.FullyQualifiedFormat)`?

Reply:
Haven't even noticed that one sticking out as slow.

This tweak was borderline premature optimisation so don't worry too much about it. The slowest things off the top of my head are formatting a file, the GetOperation API and any analysis e.g. control flow analysis.
Any micro optimisations like which order you check some already-calculated booleans are totally irrelevant compared to those big hitters so don't worry too much so long as you aren't introducing n^2 algorithms

I haven't done any profiling for a while, I should probably profile the tests some time to see if there are any quick wins, but profiling on a real code base would be even better.

Reply:
that (performance), but also following possible issues:

1. The code does checks like this:
```cs
var optionalAttribute = CommonConversions.KnownTypes.OptionalAttribute;
if (optionalAttribute == null) {
    return false;
}
```

But if `optionalAttribute` is `null`, something went terribly wrong: probably the compilation doesn't have the class for some reason... Maybe we should do something drastic in this case?

2. Things like `SymbolEqualityComparer.IncludeNullability.Equals(a.AttributeClass, defaultParameterValueAttribute)` work only if the class symbols are from the same compilation, IIRC... I hope that's always the case, but I haven't look too much into the details.